src/graph/optimizer/rule/OptimizeLeftJoinPredicateRule.cpp

```diff
@@ -0,0 +1,173 @@
+
+  auto* projectGroupNode = matched.dependencies[1].node;
+  auto* project = static_cast<graph::Project*>(projectGroupNode->node());
+
+  auto* appendVerticesGroupNode = matched.dependencies[1].dependencies[0].node;
+  auto appendVertices =
+      static_cast<graph::AppendVertices*>(matched.dependencies[1].dependencies[0].node->node());
+
+  auto traverse = static_cast<graph::Traverse*>(
+      matched.dependencies[1].dependencies[0].dependencies[0].node->node());
+
+  auto& avNodeAlias = appendVertices->nodeAlias();
+
+  auto& tvEdgeAlias = traverse->edgeAlias();
+
+  auto& hashKeys = leftJoin->hashKeys();
+  auto& probeKeys = leftJoin->probeKeys();
+
+  // Use visitor to collect all function `id` in the hashKeys
+  bool found = false;
+  size_t hashKeyIdx = 0;
+  for (size_t i = 0; i < hashKeys.size(); ++i) {
+    auto* hashKey = hashKeys[i];
+    if (hashKey->kind() != Expression::Kind::kFunctionCall) {
+      continue;
+    }
+    auto* func = static_cast<FunctionCallExpression*>(hashKey);
+    if (func->name() != "id" && func->name() != "_joinkey") {
+      continue;
+    }
+    auto& args = func->args()->args();
+    DCHECK_EQ(args.size(), 1);
+    auto* arg = args[0];
+    if (arg->kind() != Expression::Kind::kInputProperty) {
+      continue;
+    }
+    auto& alias = static_cast<InputPropertyExpression*>(arg)->prop();
+    if (alias != avNodeAlias) continue;
+    // Must check if probe keys contain the same key
+    if (*probeKeys[i] != *hashKey) {
+      return TransformResult::noTransform();
+    }
+    if (found) {
+      return TransformResult::noTransform();
+    }
+    hashKeyIdx = i;
+    found = true;
+  }
+  if (!found) {
+    return TransformResult::noTransform();
+  }
+
+  found = false;
+  size_t prjIdx = 0;
+  auto* columns = project->columns();
+  for (size_t i = 0; i < columns->size(); ++i) {
+    const auto* col = columns->columns()[i];
+    if (col->expr()->kind() != Expression::Kind::kInputProperty) {
+      continue;
+    }
+    auto* inputProp = static_cast<InputPropertyExpression*>(col->expr());
+    if (inputProp->prop() != avNodeAlias) continue;
+    if (found) {
+      return TransformResult::noTransform();
+    }
+    prjIdx = i;
+    found = true;
+  }
+  if (!found) {
+    return TransformResult::noTransform();
+  }
+
+  auto* pool = octx->qctx()->objPool();
+  // Let the new project generate expr `none_direct_dst($-.tvEdgeAlias)`,
+  // and let the new left join use it as hash key
+  auto* args = ArgumentList::make(pool);
+  args->addArgument(InputPropertyExpression::make(pool, tvEdgeAlias));
+  auto* newPrjExpr = FunctionCallExpression::make(pool, "none_direct_dst", args);
+
+  auto* newYieldColumns = pool->makeAndAdd<YieldColumns>();
+  for (size_t i = 0; i < project->columns()->size(); ++i) {
+    if (i == prjIdx) {
+      newYieldColumns->addColumn(new YieldColumn(newPrjExpr, avNodeAlias));
+    } else {
+      newYieldColumns->addColumn(project->columns()->columns()[i]->clone().release());
```

Comment:
better to use a new variable to store the result `project->columns()->columns()` to avoid to construct the column vector each time.

Reply:
Good catch!